async-nats/src/connector.rs

```diff
@@ -0,0 +1,243 @@
+use crate::connection::Connection;
+use crate::Authorization;
+use crate::ClientOp;
+use crate::Protocol;
+use crate::ServerAddr;
+use crate::ServerInfo;
+use crate::SocketAddr;
+use crate::ToServerAddrs;
+use crate::VERSION;
+use bytes::BytesMut;
+use std::cmp;
+use std::collections::HashMap;
+use std::io;
+use std::io::BufWriter;
+use std::path::PathBuf;
+use std::sync::Arc;
+use std::time::Duration;
+use tokio::io::AsyncWrite;
+use tokio::io::ErrorKind;
+use tokio::net::TcpStream;
+use tokio::time::sleep;
+use tokio_rustls::rustls::{self};
+
+pub(crate) struct ConnectorOptions {
+    pub(crate) tls_required: bool,
+    pub(crate) certificates: Vec<PathBuf>,
+    pub(crate) client_cert: Option<PathBuf>,
+    pub(crate) client_key: Option<PathBuf>,
+    pub(crate) tls_client_config: Option<rustls::ClientConfig>,
+    pub(crate) auth: Authorization,
+    pub(crate) no_echo: bool,
+}
+
+/// Maintains a list of servers and establishes connections.
+pub(crate) struct Connector {
+    /// A map of servers and number of connect attempts.
+    servers: HashMap<ServerAddr, usize>,
+    options: ConnectorOptions,
+}
+
+impl Connector {
+    pub(crate) fn new<A: ToServerAddrs>(
+        addrs: A,
+        options: ConnectorOptions,
+    ) -> Result<Connector, io::Error> {
+        let servers = addrs
+            .to_server_addrs()?
+            .into_iter()
+            .map(|addr| (addr, 0))
+            .collect();
+
+        Ok(Connector { servers, options })
+    }
+
+    pub(crate) async fn connect(&mut self) -> Result<(ServerInfo, Connection), io::Error> {
+        for _ in 0..128 {
```

Comment:
This should be settable as a Duration parameter. (in my use case I'd prefer it to run indefinitely so I would set this to a large number).
Also, do you want to wait a bit between retries? Might be another Duration parameter with default of 500ms or so.
lower priority but nice-to-have: although we do have a callback for Disconnect, where we can log the fact that the connection dropped, it would be nice to be able to log the fact that it's trying to reconnect, so I'd like to get a callback every N seconds or N retries.

Reply:
I think both me and @Jarema are in agreement with most of your points, reconnection attempts, backoff factor etc should all be configurable. We punted them to "later" when we did the first pass on reconnect, now is probably a good time to revisit these.